src/util/aws/lambda.js

```diff
@@ -1,90 +1,159 @@
-  } catch (e) {
-    if (e.code !== 'ResourceNotFoundException') { throw e }
-    const params = merge({}, config, { Publish, Code: { ZipFile } })
-    await lambda.createFunction(params).promise()
+  if (Code.s3) {
+    createFunctionParams.Code.S3Bucket = Code.s3.bucket
+    createFunctionParams.Code.S3Key = Code.s3.key
+  } else if (Code.Zip) {
+    createFunctionParams.Code.ZipFile = Code.Zip
+  } else {
+    throw new Error('need to upload code')
   }
 
-  await putEnvironment(env, config)
-  return lambda.updateFunctionCode({ ZipFile, FunctionName, Publish }).promise()
+  const newFunc = await lambda.createFunction(createFunctionParams).promise()
+
+  createAliasParams.FunctionVersion = newFunc.Version
+  await lambda.createAlias(createAliasParams).promise()
+
+  return {
+    FunctionName,
+    Identifier: {
+      Alias,
+      Version: newFunc.Version
+    },
+    Config: configStripper(newFunc),
+    Code: { CodeSha256: newFunc.CodeSha256 }
+  }
 }
 
-export async function putEnvironment (env, config, envVars) {
+// might want to check that the following are equal on these two objects:
+// FunctionName, Alias, rest are mutable
+// should have function calculate change: takes wanted function and checks it to see if there are differences?
+export async function updateFunction (oldFunction, wantedFunction) {
+  // should config should already exist
+  //validateConfig(wantedFunction.Config)
   await loadRegion()
   const lambda = new AWS.Lambda()
+  const { FunctionName, Alias } = oldFunction
+  const updateCodeParams = { FunctionName }
+  const updateConfigParams = { FunctionName }
+  const publishVersionParams = { FunctionName, Description: `Based off of ${Alias}` }
+  const updateAliasParams = { FunctionName }
+
+  if (wantedFunction.Code.Zip) {
+    updateCodeParams.ZipFile = wantedFunction.Code.Zip
+  } else if (wantedFunction.Code.s3) {
+    updateCodeParams.S3Bucket = wantedFunction.Code.s3.bucket
+    updateCodeParams.S3Key = wantedFunction.Code.s3.key
+  } else {
+    const { Code, Configuration } = await lambda.getFunction({ FunctionName: oldFunction.FunctionName, Qualifier: oldFunction.Alias }).promise()
 
-  validateConfig(config)
-
-  const params = {
-    FunctionName: config.FunctionName,
-    Qualifier: env
+    publishVersionParams.CodeSha256 = Configuration.CodeSha256
+    updateCodeParams.ZipFile = (await got(Code.Location, { encoding: null })).body
   }
+  const uploadedFunc = await lambda.updateFunctionCode(updateCodeParams).promise()
+  publishVersionParams.CodeSha256 = publishVersionParams.CodeSha256 || uploadedFunc.CodeSha256
 
-  let awsFunction
-  try {
-    awsFunction = await getFunction(params)
-  } catch (e) {
-    if (e.code !== 'ResourceNotFoundException') { throw e }
-    awsFunction = await getFunction({ FunctionName: params.FunctionName })
+  merge(updateConfigParams, configStripper(wantedFunction.Config))
+  await lambda.updateFunctionConfiguration((updateConfigParams)).promise()
```

Comment:
This api call doesn't let you supply `CodeSha256` as a parameter, but we get it in the return value. I think we should check it here ourselves to be safe.